include/algorithms/synthesis/dd_synthesis.hpp

```diff
@@ -1,9 +1,56 @@
 #pragma once
 
+#include "QuantumComputation.hpp"
 #include "core/truthTable/truth_table.hpp"
+#include "dd/FunctionalityConstruction.hpp"
+
+#include <chrono>
 
 namespace syrec {
 
     auto buildDD(const TruthTable& tt, std::unique_ptr<dd::Package<>>& dd) -> dd::mEdge;
 
+    class DDSynthesizer {
+    public:
+        DDSynthesizer() = default;
+
+        explicit DDSynthesizer(std::size_t nqubits):
+            qc(nqubits) {}
+
+        auto pathFromSrcDst(dd::mNode* src, dd::mNode const* dst, TruthTable::Cube::Vector& sigVec, TruthTable::Cube& tempVec) const -> void;
+
+        auto pathSignature(dd::mNode* src, TruthTable::Cube::Vector& sigVec, TruthTable::Cube& tempVec) -> void;
+
+        static auto terminate(dd::mEdge const& current) -> bool;
+
+        auto unifyPath(dd::mEdge& src, dd::mEdge const& current, TruthTable::Cube::Vector const& p1SigVec, TruthTable::Cube::Vector const& p2SigVec, const std::vector<std::size_t>& indices, std::unique_ptr<dd::Package<>>& dd) -> bool;
+
+        auto swapPaths(dd::mEdge& src, dd::mEdge const& current, std::unique_ptr<dd::Package<>>& dd) -> bool;
+
+        auto shiftUniquePaths(dd::mEdge& src, dd::mEdge const& current, std::unique_ptr<dd::Package<>>& dd) -> bool;
+
+        auto shiftingPaths(dd::mEdge& src, dd::mEdge const& current, std::unique_ptr<dd::Package<>>& dd) -> bool;
+
+        auto synthesize(dd::mEdge& src, std::unique_ptr<dd::Package<>>& dd) -> void;
```

Comment:
The `src` should probably not be passed by reference in order to avoid modifying it.

Reply:
So the issue is I use the synthesis function recursively inside the other function. That is the reason I have kept src to be a reference. Is there anything else that we can do?

Reply:
For cases like these the best solution probably is to create two separate functions:
- one synthesis function that takes a DD `mEdge` as a const reference and returns a quantum circuit. (this is the public interface function)
- one synthesis function that takes a DD `mEdge` as const reference and returns an `mEdge` (the result of the recursive call).
Note that these functions need to have different names as they only differ in their return type. I'd suggest calling the high level function `synthesize` and the recursive function something like `synthesizeRec`.

In addition to the above, all private member function should take their DD parameters as const reference and return the newly computed DD (instead of being void).

Reply:
done 👍🏻